### ISSUE TITLE
proposal: no-class-fields

### ISSUE THREAD

**Issue by trusktr** [eslint/eslint#12664]
**Please describe what the rule should do:**

It should disallow class fields:

```js
  class Foo {
    static #foo = 123 // should be a lint error
    #foo = 123 // should be a lint error
    foo = 123 // should be a lint error
  }
```

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

see above

**Why should this rule be included in ESLint (instead of a plugin)?**

There are many reasons not to like the new class fields, and to avoid using them. More information can be uncovered here: https://github.com/tc39/proposal-class-fields/issues/100. There are more issues not linked there, for example here's a good one on inconsistency of class fields, https://github.com/tc39/proposal-class-fields/issues/265, or inconsistency of inheritance, https://github.com/tc39/proposal-class-fields/issues/280.

We can avoid all the problems and by discouraging use of class fields. An `eslint-disable` comment can be placed in areas where it is absolutely necessary.

**Comment by ljharb** [comment]
This can already be achieved with `no-restricted-syntax`; a rule that serves no purpose except to ban an entire (very useful) language feature is not generally one that's included in eslint.

Separately, class fields is stage 3, and eslint sadly doesn't support anything prior to stage 4.

**Comment by kaicataldo** [comment]
To reiterate what @ljharb said, ESLint doesn't support syntax until it reaches stage 4. If and when that occurs, the team would be happy to consider a proposal regarding class fields.

**Comment by ilyavolodin** [comment]
Just to re-iterate on what @ljharb said, ESLint tries to avoid rules that bans language features. We have a general `no-restricted-syntax` rule that works as catch-all for banning language features, and we do not want to implement separate rules per specific feature. Even when class fields syntax makes it into Stage 4, this proposal is very unlikely to be accepted.
Thanks for the suggestion, but I'm going to close this issue since it's not actionable right now.

**Comment by dandv** [comment]
@trusktr: how did you avoid the "Unexpected token =" error that you get for static class members [by default](https://eslint.org/demo#eyJ0ZXh0IjoiY2xhc3MgRm9vIHtcbiAgc3RhdGljIGJhciA9IDEyM1xufVxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19) in the first place?
